plugins/cnfreg/discovery.go

```diff
@@ -50,51 +51,91 @@ type PidFile struct {
 	HttpPort  int    `json:"http-port"`
 }
 
-// cnfDiscovery discovers all CNFs that should be loaded into StoneWork (as StoneWork modules).
-// Currently this is a very simple procedure - StoneWork will just wait configured amount of time and each CNF should
-// in the meantime write pid file under the filepath /run/stonework/discovery/<cnf-name>.pid, containing pid, IP address,
-// gRPC and http port numbers.
-// StoneWork will then connect to each CNF over gRPC and discovers all CNF-specific configuration models.
-// Downside of this trivial approach for discovery is that the Init phase will take longer and most of that time nothing
-// actually happens (agent sleeps).
-func (p *Plugin) cnfDiscovery() error {
-	// Give CNFs some time to write pid files.
-	time.Sleep(p.config.CnfDiscoveryTimeout)
-
-	// Load all pid files written by SW-Module CNFs.
-	p.sw.modules = make(map[string]swModule)
-	pidFiles, err := ioutil.ReadDir(pidFileDir)
+func (p *Plugin) cnfDiscovery(done <-chan struct{}) {
+	watcher, err := fsnotify.NewWatcher()
 	if err != nil {
-		if os.IsNotExist(err) {
-			p.Log.Warnf("Directory \"%s\" does not exist, no CNFs will be loaded", pidFileDir)
-			return nil
-		}
-		p.Log.Errorf("failed to read directory with PID files: %v", err)
-		return err
+		p.Log.Errorf("failed to create file watcher: %v", err)
+		return
 	}
-	for _, pidFile := range pidFiles {
-		if !strings.HasSuffix(pidFile.Name(), pidFileExt) {
-			continue
-		}
-		content, err := ioutil.ReadFile(path.Join(pidFileDir, pidFile.Name()))
-		if err != nil {
-			p.Log.Errorf("failed to read PID file %s: %v", pidFile.Name(), err)
-			continue
+	defer func() {
+		if err := watcher.Close(); err != nil {
+			p.Log.Errorf("failed to close file watcher: %v", err)
 		}
-		var pf PidFile
-		err = json.Unmarshal(content, &pf)
-		if err != nil {
-			p.Log.Errorf("failed to parse PID file %s: %v", pidFile.Name(), err)
-			continue
-		}
-		swMod, err := p.getCnfModels(pf.IpAddress, pf.GrpcPort, pf.HttpPort)
+	}()
+
+	_ = os.Mkdir(pidFileDir, os.ModeDir)
+	pidFiles, err := os.ReadDir(pidFileDir)
+	if err != nil {
+		p.Log.Errorf("failed to read pid file directory: %v", err)
+	}
+	for _, pf := range pidFiles {
+		swMod, err := p.loadSwModFromFile(pidFileDir + "/" + pf.Name())
 		if err != nil {
-			p.Log.Errorf("failed to obtain CNF models (pid file: %v): %v", pidFile.Name(), err)
+			p.Log.Errorf("loading StoneWork module from file failed: %v", err)
 			continue
 		}
 		p.sw.modules[swMod.cnfMsLabel] = swMod
+		p.initCnfProxy(swMod)
+	}
+
+	if err := watcher.Add(pidFileDir); err != nil {
+		p.Log.Errorf("failed to add pid file directory to watch into file watcher: %v", err)
 	}
-	return nil
+
+	go p.discovery(watcher)
```

Comment:
Is this perfectly safe from the perspective the new pid file can be created at any point in the run of this code? If i for example create pid file between 'os.ReadDir(pidFileDir)' and 'watcher.Add(pidFileDir)', then i think that it may get lost (if i'm not getting the usage of fsnotify library wrong). I guess there must be some overlap when we are are already listening to filesystem changes and then get all pid files by 'os.ReadDir' (and processing them) and then handle when the watch reports something the the 'os.ReadDir' already processed.

Reply:
Good catch! The time window would be quite small, but yes such situation could arise. Should be fixed now by changing the order a bit: `os.Mkdir` call and immediately after `watcher.Add` is called. Just in case a file was created in the `pidFileDir` between these two calls, an `os.ReadDir` check is performed once after. Note that in order for this to work properly, the sw modules map must be safe for concurrent read/write, which I will add in next changes to this PR (the map has to be thread-safe anyway - this is not the only instance in the code where that is required).